qemu/tests/qsd_backup_pull.py

```diff
@@ -0,0 +1,296 @@
+                                self.full_backup_client_images[i].tag)
+
+    def inc_copyif(self):
+        for i, nbd_obj in enumerate(self.inc_backup_nbd_images):
+            backup_utils.copyif(self.params, nbd_obj.tag,
+                                self.inc_backup_client_images[i].tag,
+                                self.full_backup_bitmaps[i])
+
+    def export_full_backups(self):
+        for i, obj in enumerate(self.full_backup_nbd_objs):
+            obj.start_nbd_server()
+            obj.add_nbd_image(self.full_backups[i])
+
+    def stop_export_full_backups(self):
+        for obj in self.full_backup_nbd_objs:
+            obj.stop_export()
+
+    def export_inc_backups(self):
+        for i, obj in enumerate(self.inc_backup_nbd_objs):
+            obj.start_nbd_server()
+            obj.add_nbd_image(self.inc_backups[i])
+
+    def stop_export_inc_backups(self):
+        for obj in self.inc_backup_nbd_objs:
+            obj.stop_export()
+
+    def cancel_backup_jobs(self):
+        for job_id in self.backup_jobs:
+            arguments = {'id': job_id}
+            self.qsd.monitor.cmd('job-cancel', arguments)
+
+    def do_full_backup(self):
+        extra_options = {"sync": "none", "wait_job_complete": False}
+        backup_utils.blockdev_batch_backup(
+            self.qsd,
+            self.source_images,
+            self.full_backups,
+            self.full_backup_bitmaps,
+            **extra_options)
+        self.backup_jobs = [job['id']
+                            for job in job_utils.query_jobs(self.qsd)]
+
+    def generate_inc_files(self):
+        return list(map(self.generate_data_file, self.src_img_tags))
+
+    def add_target_data_disks(self, bktype='full'):
+        """Hot add target disk to VM with qmp monitor"""
+        for tag in self.params.objects("source_images"):
+            image_params = self.params.object_params(tag)
+            img = image_params['full_backup_image'] if bktype == 'full' else image_params['inc_backup_image']
+            disk = self.target_disk_define_by_params(self.params, img)
+            disk.hotplug(self.qsd)
+            self.trash.append(disk)
+
+    def do_incremental_backup(self):
+        extra_options = {"sync": "none",
+                         "disabled_bitmaps": self.disabled_bitmaps,
+                         "wait_job_complete": False}
+        backup_utils.blockdev_batch_backup(
+            self.qsd,
+            self.source_images,
+            self.inc_backups,
+            self.inc_backup_bitmaps,
+            **extra_options)
+        self.backup_jobs = [job['id']
+                            for job in job_utils.query_jobs(self.qsd)]
+
+    def restart_vm_with_backup_images(self):
+        """restart vm with back2 as its data disk"""
+        self.main_vm.destroy()
+        self.qsd.stop_daemon()
+        self.params['qsd_images_qsd1'] = self.params['client_image_inc']
+        add_vubp_into_boot(self.params['client_image_inc'], self.params)
```

Comment:
hi @aliang123 , I just found here we may also need to handle a list instead of client_image_inc, it's the same reason that the source_images is a list.

self.params['qsd_images_qsd1'] = ' '.join([i.tag for i in self.inc_backup_client_images])
for addr, inc in enumerate(self.inc_backup_client_images, start=15):
        add_vubp_into_boot(inc.tag, self.params, addr)